datamodel/tests/common.py

```diff
@@ -0,0 +1,224 @@
+    def setUp(self):
+        # should be ready only during tests, never during installation
+        # of addons
+        self._datamodels_registry.ready = True
+
+        @self.addCleanup
+        def notready():
+            self._datamodels_registry.ready = False
+
+
+class TransactionDatamodelCase(common.TransactionCase, DatamodelMixin):
+    """ A TransactionCase that loads all the datamodels
+
+    It it used like an usual Odoo's TransactionCase, but it ensures
+    that all the datamodels of the current addon and its dependencies
+    are loaded.
+
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super(TransactionDatamodelCase, cls).setUpClass()
+        cls.setUpDatamodel()
+
+    # pylint: disable=W8106
+    def setUp(self):
+        # resolve an inheritance issue (common.TransactionCase does not call
+        # super)
+        common.TransactionCase.setUp(self)
+        DatamodelMixin.setUp(self)
+
+
+class SavepointDatamodelCase(common.SavepointCase, DatamodelMixin):
+    """ A SavepointCase that loads all the datamodels
+
+    It is used like an usual Odoo's SavepointCase, but it ensures
+    that all the datamodels of the current addon and its dependencies
+    are loaded.
+
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super(SavepointDatamodelCase, cls).setUpClass()
+        cls.setUpDatamodel()
+
+    # pylint: disable=W8106
+    def setUp(self):
+        # resolve an inheritance issue (common.SavepointCase does not call
+        # super)
+        common.SavepointCase.setUp(self)
+        DatamodelMixin.setUp(self)
+
+
+class DatamodelRegistryCase(
+    unittest.TestCase, common.MetaCase("DummyCase", (object,), {})
+):
+    """ This test case can be used as a base for writings tests on datamodels
+
+    This test case is meant to test datamodels in a special datamodel registry,
+    where you want to have maximum control on which datamodels are loaded
+    or not, or when you want to create additional datamodels in your tests.
+
+    If you only want to *use* the datamodels of the tested addon in your tests,
+    then consider using one of:
+
+    * :class:`TransactionDatamodelCase`
+    * :class:`SavepointDatamodelCase`
+
+    This test case creates a special
+    :class:`odoo.addons.datamodel.core.DatamodelRegistry` for the purpose of
+    the tests. By default, it loads all the datamodels of the dependencies, but
+    not the datamodels of the current addon (which you have to handle
+    manually). In your tests, you can add more datamodels in 2 manners.
+
+    All the datamodels of an Odoo module::
+
+        self._load_module_datamodels('connector')
+
+    Only specific datamodels::
+
+        self._build_datamodels(MyDatamodel1, MyDatamodel2)
+
+    Note: for the lookups of the datamodels, the default datamodel
+    registry is a global registry for the database. Here, you will
+    need to explicitly pass ``self.datamodel_registry`` in the
+    """
+
+    def setUp(self):
```

Comment:
I've reworked these classes here to allow components registration on setupclass instead of setUp only https://github.com/OCA/connector/pull/387

BTW SavepointCase is going to disappear in master and TransactionCase will use setUpClass